### PR TITLE
[IMP] website: better retrieval of base URL 

### PR DESCRIPTION
**Pull request by rdeodoo**
Community: https://github.com/odoo/odoo/pull/68201
Enterprise: https://github.com/odoo/enterprise/pull/17538
Upgrade: https://github.com/odoo/upgrade/pull/2372
task-2476101

### CODE REVIEW

**Comment by rdeodoo** [comment]
@JKE-be :wave: About the `TODO`'s , I'd go for it.
I'll clean the code later once we agree on the order, do another pass on the remaining `ICP` and do the upgrade PR, but before going further, do you agree on the fallback order:
**rec.website_id > rec.company_id.website_id > get_current_website > self.env.company.website_id > ICP ?** (note that company.website_id is the search fallback, the field doesnt exist ofc, but maybe it should for cache purpose?)

**Comment by rdeodoo** [comment]
As discussed, **rec.website_id > rec.company_id.website_id > ~~get_current_website > self.env.company.website_id~~ > ICP** as we don't want to rely on the context but only on the record.

Also, not forget to clean the domain if localhost/0000 in upgrade script

**Comment by jco-odoo** [comment]
A lot of people pinged here (but I comment anyways).  It just seems weird to me that you can apply the method on just any object.  Why not the env or something?  (it is confusing because a _get_base_url of an invoice could be something entirely different)

**Comment by rdeodoo** [comment]
@jco-odoo The idea is that if a record is bound to a `website_id`, we want to use that website's URL (its domain) if set, before fallbacking on the ICP.
That idea got extended to the record's `company_id`, is a record is not bound to a website but is bound to a company, we try to detect the company's website though some heuristic (because there is no direct relation, a company might have more than one website), and if we found a suitable website, we use its domain.

The overal idea is, if you are creating an URL in a record context, and that record is somehow bound to a website, we want to use that website URL instead of the `web.base.url`.

**Comment by jco-odoo** [comment]
@rdeodoo Nice, but the idea of a base url for the invoice stays a bit double to me.  invoice.company_id._get_base_url() would not be more logical then?

**Comment by rdeodoo** [comment]
> @rdeodoo Nice, but the idea of a base url for the invoice stays a bit double to me. invoice.company_id._get_base_url() would not be more logical then?

We want to prioritize `website_id` over `company_id.website_id`, as what's coming from a website context will most likely have a `website_id` set than a `company_id` set.

But if no `website_id` is set, it will fallback on something equivalent to what you suggest.

Note that this PR introduce a mechanism to prevent mismatch between `company_id` and `website_id` on a record.

File: addons/sale/models/sale.py [hunk 1]
```diff
@@ -106,12 +106,15 @@ def get_empty_list_help(self, help):
         )
         return super(SaleOrder, self).get_empty_list_help(help)
 
+    @api.model
+    def _default_note_url(self):
+        return self.env['ir.config_parameter'].sudo().get_param('web.base.url')
```
Comment (JKE-be) [inline addons/sale/models/sale.py:111]: :/

File: addons/website/models/ir_model.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+           it is retrieve through some heuristic in its `website_id`'s compute.
+        3. Use the ICP `web.base.url` (super)
+
+        :return: the base url for this record
+        :rtype: string
+        """
+        self.ensure_one()
+        if 'website_id' in self and self.website_id.domain:
+            return self.website_id._get_http_domain()
+        if 'company_id' in self and self.company_id.website_id.domain:
```
Comment (JKE-be) [inline addons/website/models/ir_model.py:26]: _name + super()

File: addons/website/models/website.py [hunk 3]
```diff
@@ -198,10 +202,15 @@ def write(self, values):
...
 
         result = super(Website, self - public_user_to_change_websites).write(values)
+
         if 'cdn_activated' in values or 'cdn_url' in values or 'cdn_filters' in values:
             # invalidate the caches from static node at compile time
             self.env['ir.qweb'].clear_caches()
 
+        # invalidate cache for `company.website_id` to be recomputed
+        if 'sequence' in values or 'company_id' in values:
+            self.env['res.company'].invalidate_cache(fnames=['website_id'], ids=(original_company | self.company_id).ids)
```
Comment (JKE-be) [inline addons/website/models/website.py:212]: +create ? + delete ?

File: addons/website_crm_partner_assign/wizard/crm_forward_to_partner.py [hunk 4]
```diff
@@ -103,18 +103,18 @@ def action_forward(self):
...
+            lead.get_base_url(),
             self.env.cr.dbname,
-            lead_id,
+            lead.id,
             action_ref and action_ref.id or False)
         return portal_link
 
     def get_portal_url(self):
-        portal_link = "%s/?db=%s" % (self.env['ir.config_parameter'].sudo().get_param('web.base.url'), self.env.cr.dbname)
+        portal_link = "%s/?db=%s" % (self[0].get_base_url(), self.env.cr.dbname)
```
Comment (JKE-be) [inline addons/website_crm_partner_assign/wizard/crm_forward_to_partner.py:117]: ensureone

Comment (rdeodoo) [reply addons/website_crm_partner_assign/wizard/crm_forward_to_partner.py]: Done, even if I think this is not needed as we always have an object for the template IMO https://github.com/odoo-dev/odoo/blame/ee6146a871267c86484cf524000cc5e20d8adea7/addons/website_crm_partner_assign/data/mail_template_data.xml#L45
That also add back another ICP call

File: addons/website/models/website.py (hunk 3 above)
Comment (rdeodoo) [reply addons/website/models/website.py]: Done, + add complete tests on cache

File: addons/website/models/ir_model.py (hunk 2 above)
Comment (rdeodoo) [reply addons/website/models/ir_model.py:29]: Done

File: addons/sale/models/sale.py (hunk 1 above)
Comment (rdeodoo) [reply addons/sale/models/sale.py]: I can't call `get_base_url()` here as I don't have a record yet

**Comment by rdeodoo** [comment]
- Changes done
- Branches rebased, they were conflicts in both community & enterprise
- Added some tests in top of the existing ones (mainly for cache & company_id.website behavior)
- Upgrade script fully tested & working as expected

**Comment by rdeodoo** [comment]
Added a commit to allow `get_base_url()` to be called on an empty recordset

It still prevent it to be called on multiple records, as it is a bad practice and should be avoided. Indeed, the base url depends from the record, we can't juste use the first record of the recordset to compute it, as it might be completely wrong for other records.
eg: sale_order_1 for website 1 and sale_order_2 for website 2 -> can't use base url from sale order 1 to sale order 2.

Note that the "ensure one or no record" behavior has to be implemented in base also, not in website, eg module non website dependant do records.get_base_url(), everything is fine, then you install website and it will break

**Comment by Xavier-Do** [comment]
@robodoo override=ci/style

**Comment by JKE-be** [comment]
@robodoo r+ rebase-ff

File: odoo/models.py [hunk 5]
```diff
@@ -3320,7 +3320,8 @@ def get_base_url(self):
         :rtype: string
 
         """
-        self.ensure_one()
+        if len(self) > 1:
+            raise ValueError("Expected singleton or no record: %s" % self)
```
Comment (poma-odoo) [inline odoo/models.py:3324]: What is the point of checking for singleton or no record? It is important in the override in the website module, but here, I think it is an extra useless check

**Pull request closed without merging**